File: AtCoder/ABC/ABC101-150/abc126/abc126e.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
#define REP(i,b,e) for(int i=(b);i<(e);++i)
#define rep(i,e) for(int i=0;i<(e);++i)

struct UnionFind{
  vector<int> par;
  vector<int> rank;

  UnionFind(int n) : par(n, -1), rank(n, 0){}

  int find(int x){
    if(par[x] < 0) return x;
    else return par[x] = find(par[x]);
  }

  bool unite(int x, int y){
    x = find(x); y = find(y);
    if(x==y) return false;
    else if(rank[x]<rank[y]) par[x] = y;
    else if(rank[x]>rank[y]) par[y] = x;
    else par[y] = x, rank[x] += 1;
    return true;
  }

  bool same(int x, int y){
    return find(x)==find(y);
  }
};

int main() {
  int n, m;
  cin >> n >> m;
  UnionFind uf(n+1);
  int ans = 0;
  rep(i, m){
    int x, y, z;
    cin >> x >> y >> z;
    ans += uf.unite(x, y);
  }
  cout << n-ans << endl;
  return 0;
}

File: AtCoder/ABC/ABC101-150/abc126/abc126e.py
class UnionFind():

  def __init__(self, n):
    self.par = [i for i in range(n)]
    self.rank = [0] * n

  def find(self, x):
    if self.par[x] != x:
      self.par[x] = self.find(self.par[x])
    return self.par[x]

  def union(self, x, y):
    x, y = self.find(x), self.find(y)
    if x == y:
      return False
    elif self.rank[x] < self.rank[y]:
      self.par[x] = y
    elif self.rank[x] > self.rank[y]:
      self.par[y] = x
    else:
      self.par[y] = x
      self.rank[x] += 1
    return True

  def same(self, x, y):
    return self.find(x) == self.find(y)

n, m = list(map(int, input().split()))
uf = UnionFind(n+1)
ans = 0
for i in range(m):
  x, y, z = list(map(int, input().split()))
  ans += uf.union(x, y)
print(n-ans)